<a href="https://colab.research.google.com/github/axel-sirota/manage-data-pandas/blob/main/module3/ManageDataPandas_Mod3Demo1_IdentifyNaNs.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Identifying NaNs



## Prep

In this demo we will find ways of identifying `NaNs`, not a number, cells inside our Pandas DataFrames. NaNs are very important because they have a `carry over` effect, what I mean is:

In [ ]:
import numpy as np 
import pandas as pd

a = 2
b = np.nan

print(f'The result of {a} and {b} added is {a+b}')

The result of 2 and nan added is nan


Likewise it happens with multiplications:

In [ ]:
a = 2
b = np.nan

print(f'The result of {a} and {b} multiplied is {a*b}')

The result of 2 and nan multiplied is nan


Therefore it affects all of our calculations in Data Science like averages, standard deviations or more complex constructs like Neural Networks. To work on this, we will first download a public dataset of amount drinked per type and per country.

In [ ]:
%%writefile get_data.sh
if [ ! -f drinks.csv ]; then
  wget -O drinks.csv https://raw.githubusercontent.com/axel-sirota/manage-data-pandas/main/data/drinks.csv
fi

Overwriting get_data.sh


In [ ]:
!bash get_data.sh

## Loading the data and analizing it

In [ ]:
drinks =  pd.read_csv('drinks.csv')
drinks

country  beer_servings  spirit_servings  wine_servings  \
0    Afghanistan              0                0              0   
1        Albania             89              132             54   
2        Algeria             25                0             14   
3        Andorra            245              138            312   
4         Angola            217               57             45   
..           ...            ...              ...            ...   
188    Venezuela            333              100              3   
189      Vietnam            111                2              1   
190        Yemen              6                0              0   
191       Zambia             32               19              4   
192     Zimbabwe             64               18              4   

     total_litres_of_pure_alcohol continent  
0                             0.0        AS  
1                             4.9        EU  
2                             0.7        AF  
3                            12.4        EU  
4                             5.9        AF  
..                            ...       ...  
188                           7.7        SA  
189                           2.0        AS  
190                           0.1        AS  
191                           2.5        AF  
192                           4.7        AF  

[193 rows x 6 columns]

We can easily see for each country how much beer, spirit, wine and total amounts have been drank; as well as the continent. However, what happens if we describe the dataframe?


In [ ]:
drinks.describe()

beer_servings  spirit_servings  wine_servings  \
count     193.000000       193.000000     193.000000   
mean      106.160622        80.994819      49.450777   
std       101.143103        88.284312      79.697598   
min         0.000000         0.000000       0.000000   
25%        20.000000         4.000000       1.000000   
50%        76.000000        56.000000       8.000000   
75%       188.000000       128.000000      59.000000   
max       376.000000       438.000000     370.000000   

       total_litres_of_pure_alcohol  
count                    193.000000  
mean                       4.717098  
std                        3.773298  
min                        0.000000  
25%                        1.300000  
50%                        4.200000  
75%                        7.200000  
max                       14.400000

It takes away the non numeric columns, so let's include them all:

In [ ]:
drinks.describe(include='all')

country  beer_servings  spirit_servings  wine_servings  \
count           193     193.000000       193.000000     193.000000   
unique          193            NaN              NaN            NaN   
top     Afghanistan            NaN              NaN            NaN   
freq              1            NaN              NaN            NaN   
mean            NaN     106.160622        80.994819      49.450777   
std             NaN     101.143103        88.284312      79.697598   
min             NaN       0.000000         0.000000       0.000000   
25%             NaN      20.000000         4.000000       1.000000   
50%             NaN      76.000000        56.000000       8.000000   
75%             NaN     188.000000       128.000000      59.000000   
max             NaN     376.000000       438.000000     370.000000   

        total_litres_of_pure_alcohol continent  
count                     193.000000       170  
unique                           NaN         5  
top                              NaN        AF  
freq                             NaN        53  
mean                        4.717098       NaN  
std                         3.773298       NaN  
min                         0.000000       NaN  
25%                         1.300000       NaN  
50%                         4.200000       NaN  
75%                         7.200000       NaN  
max                        14.400000       NaN

It appears that the continent column has some problem since it only registers 170 rows. `.describe(include='all)` is a great method for diagnosing problems; another tool is `.info()`

In [ ]:
drinks.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 193 entries, 0 to 192
Data columns (total 6 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   country                       193 non-null    object 
 1   beer_servings                 193 non-null    int64  
 2   spirit_servings               193 non-null    int64  
 3   wine_servings                 193 non-null    int64  
 4   total_litres_of_pure_alcohol  193 non-null    float64
 5   continent                     170 non-null    object 
dtypes: float64(1), int64(3), object(2)
memory usage: 9.2+ KB


Same idea, in a much more compact form we can identify NaNs this way. But one question remains: which rows have that NaN? 

## Finding which rows have a NaN

There is a great method called `.isna()` that will return True if that value is `NaN`, let's see it.

In [ ]:
drinks.isna()

country  beer_servings  spirit_servings  wine_servings  \
0      False          False            False          False   
1      False          False            False          False   
2      False          False            False          False   
3      False          False            False          False   
4      False          False            False          False   
..       ...            ...              ...            ...   
188    False          False            False          False   
189    False          False            False          False   
190    False          False            False          False   
191    False          False            False          False   
192    False          False            False          False   

     total_litres_of_pure_alcohol  continent  
0                           False      False  
1                           False      False  
2                           False      False  
3                           False      False  
4                           False      False  
..                            ...        ...  
188                         False      False  
189                         False      False  
190                         False      False  
191                         False      False  
192                         False      False  

[193 rows x 6 columns]

Now we only need an indexer! let's do it on the column `continent` which is the one with NaNs

In [ ]:
drinks[drinks.isna()['continent']]

country  beer_servings  spirit_servings  \
5               Antigua & Barbuda            102              128   
11                        Bahamas            122              176   
14                       Barbados            143              173   
17                         Belize            263              114   
32                         Canada            240              122   
41                     Costa Rica            149               87   
43                           Cuba             93              137   
50                       Dominica             52              286   
51             Dominican Republic            193              147   
54                    El Salvador             52               69   
68                        Grenada            199              438   
69                      Guatemala             53               69   
73                          Haiti              1              326   
74                       Honduras             69               98   
84                        Jamaica             82               97   
109                        Mexico            238               68   
122                     Nicaragua             78              118   
130                        Panama            285              104   
143             St. Kitts & Nevis            194              205   
144                     St. Lucia            171              315   
145  St. Vincent & the Grenadines            120              221   
174             Trinidad & Tobago            197              156   
184                           USA            249              158   

     wine_servings  total_litres_of_pure_alcohol continent  
5               45                           4.9       NaN  
11              51                           6.3       NaN  
14              36                           6.3       NaN  
17               8                           6.8       NaN  
32             100                           8.2       NaN  
41              11                           4.4       NaN  
43               5                           4.2       NaN  
50              26                           6.6       NaN  
51               9                           6.2       NaN  
54               2                           2.2       NaN  
68              28                          11.9       NaN  
69               2                           2.2       NaN  
73               1                           5.9       NaN  
74               2                           3.0       NaN  
84               9                           3.4       NaN  
109              5                           5.5       NaN  
122              1                           3.5       NaN  
130             18                           7.2       NaN  
143             32                           7.7       NaN  
144             71                          10.1       NaN  
145             11                           6.3       NaN  
174              7                           6.4       NaN  
184             84                           8.7       NaN

There we have them! Can you find out what they have in common?

**Hint:** Try to check the file and this table and verify!